src/Hierarchy.elm

```diff
@@ -0,0 +1,267 @@
+                        ( folder.id :: h.rootNodes, childIndexWithFolder, nodeDBWithFolder )
+
+                    Just p ->
+                        ( h.rootNodes
+                        , Dict.update p (addChild folder.id) childIndexWithFolder
+                        , incrementCounts p nodeDBWithFolder
+                        )
+        in
+            Ok <|
+                Hierarchy
+                    { h
+                        | nodeDB = newNodeDB
+                        , childIndex = newChildIndex
+                        , rootNodes = newRootNodes
+                    }
+
+
+{-| Add a file to the hierarchy
+-}
+registerFile : Audience -> Hierarchy -> Result String Hierarchy
+registerFile file (Hierarchy h) =
+    if Dict.member file.id h.nodeDB then
+        Err <| "Node with ID " ++ (toString file.id) ++ " has already been added"
+    else
+        let
+            nodeDBWithFile =
+                Dict.insert file.id (File file) h.nodeDB
+
+            ( newRootNodes, newChildIndex, newNodeDB ) =
+                case file.folder of
+                    Nothing ->
+                        ( file.id :: h.rootNodes, h.childIndex, nodeDBWithFile )
+
+                    Just p ->
+                        ( h.rootNodes
+                        , Dict.update p (addChild file.id) h.childIndex
+                        , incrementCounts p nodeDBWithFile
+                        )
+        in
+            Ok <|
+                Hierarchy
+                    { h
+                        | nodeDB = newNodeDB
+                        , childIndex = newChildIndex
+                        , rootNodes = newRootNodes
+                    }
+
+
+{-| Adds a node ID to the list of node references
+This is a helper function used in a `Dict.update` call applied to the
+children index.
+-}
+addChild : Int -> Maybe (List Int) -> Maybe (List Int)
+addChild id maybeList =
+    maybeList
+        |> Maybe.map ((::) id)
+        |> Maybe.orElseLazy (\_ -> Just [ id ])
+
+
+{-| We assume the count is computed as the amount of
+   DIRECT children nodes.
+-}
+incrementCounts : Int -> Dict Int Node -> Dict Int Node
+incrementCounts id nodeDB =
+    let
+        updateFolderCount mf =
+            case mf of
+                Just (Folder c af) ->
+                    Just (Folder (c + 1) af)
+
+                Nothing ->
+                    mf
+
+                Just file ->
+                    -- TODO: better error handling
+                    Debug.crash "uh, somebody claimed to be the child of a file"
+    in
+        Dict.update id updateFolderCount nodeDB
+
+
+{-| Ordering for nodes
+Alphabetical, but folders come first
+-}
+nodeCompare : Node -> Node -> Order
+nodeCompare a b =
+    case ( a, b ) of
+        ( File x, File y ) ->
+            compare x.name y.name
+
+        ( Folder _ x, Folder _ y ) ->
+            compare x.name y.name
+
+        ( File _, Folder _ _ ) ->
+            GT
+
+        ( Folder _ _, File _ ) ->
+            LT
```

Comment:
This solution is not bad but I think it show that it's not that trivial and sometimes choosing simple solution can quite explode in a way you start handling cases all over the place. Personally I would abstract from the thing little bit, solve just that abstract problem and then reused this abstraction. I've built small package for solving hierarchy navigation abstractly if you're interested. It's what we are using: https://github.com/turboMaCk/chae-tree

It's really nice that you separated this from `Model, Update, View` logic but I would push it one level of abstraction up.

Hard to say what solution is better. Personally I like to break problem into smaller ones even though as a whole it can lead to slightly bigger solution it's bit more maintainable in my experience since business logic is not missed up with algorithms.

Anyway I'm giving you 👍 👍 👍. Good job in dealing with this. This implementation is indeed one of the best I saw for similar problems.